### PR TITLE
fix: input validation in critical system

### LINKED ISSUE #12: Missing Input Validation in Critical Systems
# Missing Input Validation in Critical Systems

## Description

Several systems lack proper input validation, which could lead to unexpected behavior, crashes, or exploits. User input and inter-system communication should always be validated.

### storageSystem.js
```javascript
depositFromInventory(categoryOrId, itemIdOrQty, quantity = 1) {
    // Only sanitizes quantity with Math.floor
    let qty = Math.floor(quantity);

    // No validation for negative quantities until later
    if (qty < 1) return false;

    // itemId could be any value
}
```

### currencyManager.js
```javascript
earn(amount, source = "unknown") {
    if (amount <= 0) return false;  // Good but...
    // No check for NaN, Infinity, or non-numbers
    this.currentMoney += amount;
}

spend(amount, item = "unknown") {
    if (amount <= 0) return false;
    // Same issue - could pass NaN
}
```

### inventorySystem.js
```javascript
addItem(categoryOrId, itemIdOrQty, quantity = 1, _recursionDepth = 0) {
    // Good recursion protection but...
    // No validation that itemId is a valid number
    // No validation that quantity is a positive integer
}
```

### merchant.js (from preview)
```javascript
// Trade quantities could potentially be manipulated
// Price calculations should be validated
```

### 1. Create validation utilities

```javascript
// scripts/validation.js
export function isValidPositiveInteger(value) {
    return Number.isInteger(value) && value > 0;
}

export function isValidPositiveNumber(value) {
    return typeof value === 'number' &&
           !isNaN(value) &&
           isFinite(value) &&
           value > 0;
}

export function isValidItemId(id, items) {
    return isValidPositiveInteger(id) &&
           items.some(item => item.id === id);
}

export function sanitizeQuantity(value, min = 1, max = 9999) {
    if (typeof value !== 'number' || isNaN(value)) return min;
    return Math.max(min, Math.min(max, Math.floor(value)));
}

export function validateTradeInput(amount, maxAvailable) {
    if (!isValidPositiveInteger(amount)) {
        return { valid: false, error: 'Invalid amount' };
    }
    if (amount > maxAvailable) {
        return { valid: false, error: 'Insufficient quantity' };
    }
    return { valid: true };
}
```

### 2. Apply validation in CurrencyManager

```javascript
import { isValidPositiveNumber } from './validation.js';

earn(amount, source = "unknown") {
    if (!isValidPositiveNumber(amount)) {
        console.warn('Invalid earn amount:', amount);
        return false;
    }

    // Prevent overflow
    const MAX_MONEY = Number.MAX_SAFE_INTEGER;
    if (this.currentMoney + amount > MAX_MONEY) {
        this.currentMoney = MAX_MONEY;
    } else {
        this.currentMoney += amount;
    }

    // ... rest of method
}
```

### 3. Apply validation in InventorySystem

```javascript
import { isValidPositiveInteger, isValidItemId, sanitizeQuantity } from './validation.js';

addItem(categoryOrId, itemIdOrQty, quantity = 1, _recursionDepth = 0) {
    // Validate recursion depth
    if (_recursionDepth > 100) {
        console.error('Recursion limit exceeded');
        return false;
    }

    // Normalize and validate inputs
    let id, qty;

    if (typeof categoryOrId === 'number') {
        id = categoryOrId;
        qty = sanitizeQuantity(itemIdOrQty);
    } else {
        id = itemIdOrQty;
        qty = sanitizeQuantity(quantity);
    }

    // Validate item exists
    if (!isValidItemId(id, items)) {
        console.error(`Invalid item ID: ${id}`);
        return false;
    }

    // ... rest of method
}
```

### 4. Validate merchant transactions

```javascript
processTrade(merchantId, itemId, quantity, tradeType) {
    // Validate all inputs
    if (!this.merchants.has(merchantId)) {
        return { success: false, error: 'Invalid merchant' };
    }

    if (!isValidPositiveInteger(quantity)) {
        return { success: false, error: 'Invalid quantity' };
    }

    if (!['buy', 'sell'].includes(tradeType)) {
        return { success: false, error: 'Invalid trade type' };
    }

    // Validate price hasn't been tampered with
    const expectedPrice = this.calculatePrice(itemId, quantity, tradeType);
    // ... continue with validated data
}
```

## Labels

- security
- bug
- medium-priority

### PR DESCRIPTION
**Pull request by Gabriel-Andraade**
fix  #12 

- adds validation utilities (positive integer/finite number/sanitizeQuantity/itemId checks)
- hardens CurrencyManager against NaN/Infinity and unsafe values
- validates inventory/storage transfers (itemId + qty) to avoid state corruption
- validates merchant trades (qty + mode) and recalculates expected price server-side (no trust in UI)

goal: prevent crashes, unexpected behavior, and exploit paths caused by missing input validation.

## Summary by CodeRabbit

* **New Features**
  * Centralized validation utilities and unified price calculation for trades.
  * Player needs restore now accepts defaulted parameters for safer restores.

* **Bug Fixes**
  * Comprehensive input validation and sanitization across currency, trading, storage, and inventory to prevent invalid inputs, overflow, and incorrect balances.
  * Improved trade flow error handling, affordability checks, and rollback behavior.

* **Tests**
  * Added a browser validation test suite covering currency, inventory, storage, trading, and needs edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: public/scripts/validation.js [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+ * CRITICAL: Blocks NaN and Infinity which bypass normal comparisons
+ *
+ * @param {any} value - Value to check
+ * @returns {boolean} Whether the value is a valid positive number
+ *
+ * @example
+ * isValidPositiveNumber(5)        // true
+ * isValidPositiveNumber(5.5)      // true
+ * isValidPositiveNumber(0)        // false
+ * isValidPositiveNumber(-1)       // false
```
Comment (cirospaciari) [inline public/scripts/validation.js]: **Bug: Inconsistent type coercion policy**

`validateRange` uses `Number(value)` which silently coerces strings like `"50"` → `50`. This is inconsistent with `sanitizeQuantity` and `isValidPositiveNumber` which use strict `typeof value !== 'number'` checks to reject strings.

```js
// Current (inconsistent):
const num = Number(value);

// Should be (consistent with other validators):
if (typeof value !== 'number') return min;
const num = value;
```

File: public/scripts/merchant.js [hunk 2]
```diff
@@ -859,8 +875,33 @@ class MerchantSystem {
...
+            this.showMessage('Modo de transação inválido', 'error');
+            this.hideConfirmModal();
+            return;
+        }
+
+        // ✅ CRÍTICO: Validar tradeQuantity
+        if (!isValidPositiveInteger(this.tradeQuantity)) {
+            logger.error('[Merchant] Invalid quantity:', this.tradeQuantity);
+            this.showMessage('Quantidade inválida', 'error');
+            this.hideConfirmModal();
```
Comment (cirospaciari) [inline public/scripts/merchant.js]: **Bug: `0` used as both error value and valid price**

Returning `0` when an item isn't found means a legitimate free item (price = 0) would also be caught by `isValidPositiveNumber(expectedPrice)` downstream and incorrectly rejected.

```js
// Fix: return null for errors
if (!this.selectedPlayerItem) return null;
const itemData = getItem(this.selectedPlayerItem);
if (!itemData) return null;
```

Then in callers:
```js
if (expectedPrice === null) { /* error */ }
```

File: public/scripts/storageSystem.js [hunk 3]
```diff
@@ -191,18 +192,28 @@ export class StorageSystem {
 
     let inventoryCategory = categoryOrId;
     let itemId = itemIdOrQty;
```
Comment (cirospaciari) [inline public/scripts/storageSystem.js]: **Bug: `sanitizeQuantity(NaN)` silently deposits 1 item**

`sanitizeQuantity(NaN)` returns `1` (the min), so passing `NaN` as quantity will silently deposit/withdraw 1 item instead of failing. This is silent data mutation.

Add an explicit rejection before sanitizing:
```js
if (typeof quantity !== 'number' || !Number.isFinite(quantity)) {
  console.warn('[Storage] Invalid quantity:', quantity);
  return false;
}
const qty = sanitizeQuantity(quantity, 1, 9999); // now only clamps valid numbers
```

File: public/scripts/thePlayer/playerSystem.js [hunk 4]
```diff
@@ -178,16 +187,35 @@ export class PlayerSystem {
...
-        this.needs.hunger = Math.max(0, this.needs.hunger - (rates.hunger * multiplier));
-        this.needs.thirst = Math.max(0, this.needs.thirst - (rates.thirst * multiplier));
-        
-        this.needs.energy = Math.max(0, Math.min(100, 
-            this.needs.energy - (rates.energy * multiplier)));
-        
+
+        // ✅ Aplicar com validação (clamps to 0-100)
+        this.needs.hunger = validateRange(
+            this.needs.hunger - (rates.hunger * validMultiplier),
```
Comment (cirospaciari) [inline public/scripts/thePlayer/playerSystem.js]: **Nit: Dead code — `hunger !== undefined` is always true**

Since the function signature uses default parameters `(hunger = 0, thirst = 0, energy = 0)`, these values are never `undefined`. The `!== undefined` checks are dead code.

```js
// Current (redundant):
if (hunger !== undefined && hunger !== 0) {

// Should be:
if (hunger !== 0) {
```

File: public/scripts/validationTests.js [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+import { inventorySystem } from './thePlayer/inventorySystem.js';
+import { playerSystem } from './thePlayer/playerSystem.js';
+
+/**
+ * Runs all validation tests
+ * @returns {Object} Test results summary
+ */
+export function runValidationTests() {
+    console.group('🧪 VALIDATION TEST SUITE');
+
```
Comment (cirospaciari) [inline public/scripts/validationTests.js]: **Design: Tests mutate live game state**

These tests directly manipulate `currencyManager.currentMoney`, `playerSystem.needs`, etc. on the live game instance. The `finally` blocks help, but if an exception occurs in an unexpected place, game state is corrupted.

Consider:
- Creating isolated instances: `new CurrencyManager()` for each test
- Migrating to a proper test framework (Vitest/Jest) with `beforeEach`/`afterEach`
- At minimum, snapshot and fully restore all state in a top-level try/finally

**Review by cirospaciari — COMMENTED** [review]
## Code Review Summary

This PR adds meaningful input validation to prevent NaN/Infinity exploits — good intent and important work. However, there are several issues across the implementation that should be addressed before merging. See inline comments for specifics.

---

### AI Agent Fix Prompt

Use the following prompt with an AI coding agent (Claude, Copilot, etc.) to address all issues found in this review:

> You are fixing PR #37 on `Gabriel-Andraade/FarmXP_Browser`. Apply these fixes:
>
> 1. **`public/scripts/validation.js` — `validateRange()` inconsistency**: Replace `Number(value)` coercion with a strict `typeof value !== 'number'` guard to match the strict policy in `sanitizeQuantity` and `isValidPositiveNumber`. Strings like `"50"` must be rejected, not silently coerced.
>
> 2. **`public/scripts/validation.js` — `MAX_CURRENCY` too high**: Change from `Number.MAX_SAFE_INTEGER` to a game-appropriate cap like `1_000_000_000`.
>
> 3. **`public/scripts/thePlayer/inventorySystem.js` — `itemData` variable shadowing in `addItem()`**: The new validation block declares `const itemData = getItem(id)` but the original code further down already declares `itemData`. Remove the duplicate — either reuse the first result or move all validation before the original declaration.
>
> 4. **`public/scripts/merchant.js` — `calculateExpectedPrice()` returns `0` for errors**: Return `null` instead of `0` for error cases, and update callers to check for `null`. A zero-price item is valid; using `0` as an error sentinel blocks legitimate free items.
>
> 5. **`public/scripts/storageSystem.js` — `sanitizeQuantity(NaN)` silently becomes `1`**: In `depositFromInventory`, `withdrawToInventory`, `addItem`, and `removeItem`, add an explicit guard before `sanitizeQuantity`: `if (typeof quantity !== 'number' || !Number.isFinite(quantity)) return false;`. Use `sanitizeQuantity` only for clamping already-valid numbers.
>
> 6. **`public/scripts/thePlayer/playerSystem.js` — Dead code in `restoreNeeds`**: Remove `hunger !== undefined`, `thirst !== undefined`, `energy !== undefined` checks. Default parameters `(hunger = 0, ...)` guarantee they're never undefined. Keep only the `!== 0` check.
>
> 7. **`public/scripts/validationTests.js` — Tests mutate live game state**: Refactor to create isolated instances or migrate to a proper test framework (Vitest/Jest) with setup/teardown.
>
> 8. **General — Remove noisy inline comments**: Remove all `// ✅ CRÍTICO:` and `// ✅ Validar` comments. The function names are self-documenting. Keep only comments that explain *why*.
>
> After applying, verify: no variable shadowing errors, NaN/Infinity inputs are rejected (not silently coerced to valid values), and all tests pass.
